KTrussButtJoint

ericgozzi opened the pull request:
Joint of K topology, the two main beams are bot chutted with a `DoubleCut` feature whille the cross beam with a `Pocket` feature. The common cutting plane for the two main beams is build along the bisector line between the two main beams.

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas_timber.datastructures.Beam`.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)




<img width="2242" height="1191" alt="k-truss-buut" src="https://github.com/user-attachments/assets/dac77486-b0c5-4e5e-a546-20e6453ed0fa" />

[GH.zip](https://github.com/user-attachments/files/23820895/GH.zip)

obucklin commented on src/compas_timber/connections/k_truss_butt.py:
```diff
@@ -0,0 +1,366 @@
+
+        # Create rotation plane of the bisector
+        rotation_plane = Plane.from_point_and_two_vectors(intersection_point, dir1, dir2)
+
```
this is probably faster if you take cross vectors of the two beams and get cross vector of that with the bisector direction.

ericgozzi commented on src/compas_timber/connections/k_butt.py:
```diff
@@ -0,0 +1,412 @@
+        elif cls.is_cross_beam(beam3, beam1, beam2):
+            return beam3, [beam1, beam2]
+        else:
+            raise ValueError("Could not identify cross beam in K configuration")
```
@obucklin and @papachap  Would it make sense to have a gneric/abstract parent class `KJoint` that contains these two methods and similars? These two methods could be needed for a lot of other Joints in this topology. 
In this commit I am calling this two methods also from the `KMiterJoint`.

ericgozzi replied:
Thank's a lot @obucklin !
With this method `Cluster.parse_main_and_cross_beams()`  and a override of `promote_cluster` in each KJoint it works perfectly also with the rules.

obucklin commented on src/compas_timber/connections/k_miter.py:
```diff
@@ -0,0 +1,342 @@
+        if len(cluster.joints) <= 2 or len(cluster.joints) >= 4:  # not a K_TOPO
+            raise BeamJoiningError("Cluster does not represent a K topology joint.")
+        main_beams, cross_beams = cluster.parse_main_and_cross_beams()
+        elements = list(cross_beams) + list(main_beams)
```
maybe check that there is only one cross beam.

ericgozzi commented:
Hello!

Following the concerns about creating _“throw-away”_ joints and the suggestion to expose the relevant methods externally, I came up with the following proposal.
At the moment, this is implemented only for the ButtJoint class.

Instead of instantiating a full joint object, the idea is to directly call the meaningful operation, for example:

```
feature = ButtJoint.cut_main_beam(
    self.cross_beam,
    self.main_beam,
    mill_depth=self.mill_depth
)
self.features.append(feature)
```

This way, the joint class acts also as _"feature generator"_ and must not be fully instantiated.


@obucklin , @chenkasirer , @papachap Before moving forward with this approach, I would really appreciate hearing your thoughts.

ericgozzi commented:
> @ericgozzi hey whats the status of this. It would be good if you could work on addressing all the comments and then we can get this merged.

@obucklin  It is ready for a new round of review 😊

obucklin commented on src/compas_timber/connections/butt_joint.py:
```diff
@@ -219,35 +210,80 @@ def add_features(self) -> None:
+            return lap_feature
+
+    @staticmethod
+    def pocket_on_cross_beam(cross_beam: Beam, main_beam: Beam, mill_depth: Optional[float] = None, butt_plane: Optional[Plane] = None, conical_tool: bool = False) -> Pocket:
```
Make this a bit clearer. something like `get_cross_beam_pocket_feature`

obucklin commented on src/compas_timber/connections/butt_joint.py:
```diff
@@ -219,35 +210,80 @@ def add_features(self) -> None:
+        self.cross_beam = model[self.cross_beam_guid]
+
+    @staticmethod
+    def cut_main_beam(cross_beam: Beam, main_beam: Beam, mill_depth: Optional[float] = None, butt_plane: Optional[Plane] = None) -> JackRafterCutProxy:
```
Make this a bit clearer. something like `get_main_beam_jack_cut_feature`

obucklin reviewed (changes requested):
Hey @ericgozzi. I looked at this and made some comments. I think that your approach of changing to static methods is great, but also consider if things can be made into functions, too. I think this is part of a much larger refactoring, but it's good to get a start here as a test case. 

Please also add your new joints to the class_diagrams.rst doc.

papachap commented on src/compas_timber/connections/joint.py:
```diff
@@ -8,6 +11,11 @@
+if TYPE_CHECKING:
+    from compas.geometry import Vector
+
+    from compas_timber.elements.beam import Beam
```
please use second level imports (ie. `from compas_timber.elements import Beam`) as this is the import style followed in the compas ecosystem
